common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,490 @@
+     * @return null in either case.
+     */
     @Override
     public Void processChallenge(@NonNull final ClientCertRequest request) {
         //final String methodTag = TAG + ":processChallenge";
-        return handleOnDeviceCertAuth(request);
+        synchronized (deviceLock) {
+            if (mDevice != null) {
+                handleSmartcardCertAuth(request);
+                return null;
+            }
+        }
+        //Else, proceed with user certificates stored on device.
+        handleOnDeviceCertAuth(request);
+        return null;
+    }
+
+    /**
+     * Collects YubiKitCertDetails from PIV certificates on YubiKey.
+     * If certificates are found on YubiKey, the method proceeds with the smartcard certificate picker.
+     * Otherwise, appropriate error dialogs are shown.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     */
+    private void handleSmartcardCertAuth(@NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + ":handleSmartcardCertAuth";
+        //A connection to the YubiKey needs to be made in order to read the certificates off it.
+        mDevice.requestConnection(UsbSmartCardConnection.class, new Callback<Result<UsbSmartCardConnection, IOException>>() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void invoke(@NonNull Result<UsbSmartCardConnection, IOException> value) {
+                try {
+                    final SmartCardConnection c = value.getValue();
+                    PivSession piv = new PivSession(c);
+                    //Check if too many PIN attempts have been made.
+                    if (piv.getPinAttempts() == 0) {
+                        showMaxAttemptsDialog();
+                        request.cancel();
+                        Logger.info(methodTag,  "User has reached the maximum failed attempts allowed.");
+                        return;
+                    }
+                    //Create List that contains cert details only pertinent to the cert picker.
+                    final List<YubiKitCertDetails> certList = getCertDetailsFromKey(piv);
+                    //If no certs were found, cancel flow.
+                    if (certList.isEmpty()) {
+                        request.cancel();
+                        Logger.info(methodTag,  "No PIV certificates found on YubiKey device.");
+                        //TODO: show error dialog stating that no PIV certificates were found on the YubiKey.
+                        return;
+                    }
+                    //Build and show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
+                    showSmartcardCertPickerDialog(certList, request);
+
+                } catch(final IOException e) {
+                    Logger.error(methodTag, "IOException", e);
```

Comment:
Do we want to log `e.getMessage` in the message here and below?

Reply:
oh, I incorrectly assumed that Logger would automatically log `e.getMessage()` as well. I'll add `e.getMessage()` to all error log statements.